src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -22,15 +22,52 @@ internal partial class Binder
             BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
         {
             // PROTOTYPE(tuple-equality) Block in expression tree
-
             TupleBinaryOperatorInfo.Multiple operators = BindTupleBinaryOperatorNestedInfo(node, kind, left, right, diagnostics);
 
-            // PROTOTYPE(tuple-equality) We'll save the converted nodes separately, for the semantic model
-            //BoundExpression convertedLeft = GenerateConversionForAssignment(operators.LeftConvertedType, left, diagnostics);
-            //BoundExpression convertedRight = GenerateConversionForAssignment(operators.RightConvertedType, right, diagnostics);
+            DiagnosticBag discardDiagnostics = DiagnosticBag.GetInstance();
+
+            // The converted types are only used for the semantic model, so we don't need the conversion diagnostics
+            BoundExpression convertedLeft = ApplyConvertedTypes(left, operators, isRight: false, discardDiagnostics);
+            BoundExpression convertedRight = ApplyConvertedTypes(right, operators, isRight: true, discardDiagnostics);
+
+            discardDiagnostics.Free();
 
             TypeSymbol resultType = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
-            return new BoundTupleBinaryOperator(node, left, right, kind, operators, resultType);
+
+            return new BoundTupleBinaryOperator(node, left, right, convertedLeft, convertedRight, kind, operators, resultType);
+        }
+
+        BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOperatorInfo @operator, bool isRight, DiagnosticBag diagnostics)
+        {
+            TypeSymbol convertedType = isRight ? @operator.RightConvertedTypeOpt : @operator.LeftConvertedTypeOpt;
+
+            if (@operator.InfoKind == TupleBinaryOperatorInfo.KindEnum.Multiple && expr.Kind == BoundKind.TupleLiteral)
+            {
+                var multiple = (TupleBinaryOperatorInfo.Multiple)@operator;
+                var tuple = (BoundTupleLiteral)expr;
+                ImmutableArray<BoundExpression> arguments = tuple.Arguments;
+                int length = arguments.Length;
+
+                Debug.Assert(length == multiple.Operators.Length);
```

Comment:
>Debug.Assert(length == multiple.Operators.Length); [](start = 16, length = 50)

What about error cases when the ```multiple.Operators``` is empty? How do we handle cases like that?

Reply:
Yes. That is handled in the next PR (with test hitting it). 
Let me know if I should move the fix into this PR instead.

Reply:
> Let me know if I should move the fix into this PR instead.

I think that would be the right thing to do, especially if this wouldn't be the only follow-up change in this PR.

---
In reply to: [173259632](https://github.com/dotnet/roslyn/pull/25297#discussion_r173259632) [](ancestors = 173259632)